# Lab 1 Exercises

## 1. DNA exercise
a. Read and then run the python program “dna.py”, what does it do? Try to understand it
first. Then run it.

This code snippet takes a DNA sequence as input, counts the occurrences of "c" and "g" characters in the sequence, and calculates and displays the percentage of C and G in the DNA sequence.

In [1]:
dna = str.lower(input("What is your DNA? (enter a sequence of a t c g) "))
countCG = 0
for x in dna:
	if x == "c" or x == "g":
		countCG = countCG + 1

percent = 100 * countCG / len(dna)
print ("Your DNA consists of", percent, "% C and G.")

What is your DNA? (enter a sequence of a t c g)  abcg


Your DNA consists of 50.0 % C and G.


## b. Extend the program by printing statistics for the ‘t’ and ‘a’ bases.
If the input to the program is a sequence like: ‘aatcgac’, it should print in its output the following: aat, atc, tcg, cga. That is all triplets that are produced by a sliding window of length 3 (it should work for sliding windows of different length also). E.g.: [aat]cgac a[atc]gac aa[tcg]ac aat[cga]c. Hint: you will need to know how to access individual elements of a string.

In [66]:
# Input DNA sequence
dna = input("What is your DNA? (enter a sequence of a t c g) ").lower()

# Initialize counters for C+G and T+A
countCG = 0
countTA = 0

# Count the occurrences of C+G and T+A in the DNA sequence
for x in dna:
    if x == "c" or x == "g":
        countCG += 1
    elif x == "t" or x == "a":
        countTA += 1

# Calculate the total number of bases in the DNA sequence
total_bases = len(dna)

# Calculate the percentages of C+G and T+A
percentCG = 100 * countCG / total_bases
percentTA = 100 * countTA / total_bases

# Print the percentages
print("Your DNA consists of {:.2f}% C and G.".format(percentCG))
print("Your DNA consists of {:.2f}% T and A.".format(percentTA))

# Generate and print all triplets produced by a sliding window of length 3
triplets = [dna[i:i+3] for i in range(len(dna)-2)]

for i, triplet in enumerate(triplets):
    print(f"{triplet} [{dna[i:i+3]}]{dna[i+3:]}")

What is your DNA? (enter a sequence of a t c g)  aatcgac


Your DNA consists of 42.86% C and G.
Your DNA consists of 57.14% T and A.
aat [aat]cgac
atc [atc]gac
tcg [tcg]ac
cga [cga]c
gac [gac]


## c. Try sampling the input: i.e. the input is a sequence: ‘aatcgac...’ the output should be a shorter sequence, which is produced by sampling each input element with probability ‘p’. (Hint: Try: random.rand())

In [77]:
import random

def sample_sequence(sequence, p):
  # Create an empty list to store the sampled sequence.
  sampled_sequence = []

  # Iterate over the input sequence.
  for element in sequence:

    # Generate a random number between 0 and 1.
    random_number = random.random()

    # If the random number is less than the probability `p`, add the element to
    # the sampled sequence.
    if random_number < p:
      sampled_sequence.append(element)

  # Return the sampled sequence.
  return sampled_sequence


# Example usage:

sequence = "aatcgacatcgattcggaatgac"
p = 0.5

sampled_sequence = sample_sequence(sequence, p)

print(sampled_sequence)

['a', 'a', 'c', 'g', 'g', 'a', 'g', 'a']


## 2. squares.py
a. Figure out what it does

In [44]:
# Initialize a variable 'sum' to store the sum of squares.
sum = 0

# Iterate through numbers from 1 to 10 (inclusive).
for i in range(1, 11):
    # Calculate the square of the current number 'i' and add it to the 'sum'.
    sum = sum + (i * i)

# Print the sum of the first 10 squares.
print("sum of first 10 squares is", sum)

sum of first 10 squares is 385


This code snippet calculates the sum of the squares of the first 10 positive integers (from 1 to 10) and then prints the result. In this case, the sum of the squares from 1^2 to 10^2, which is 385.

## b. Then write a function called ‘prod’ in a new programme called ‘product2.py’. This will find the products of numbers i.e. 1*2*3* ...* n. User will enter the value of ‘n’

In [45]:
# Define a function 'prod' to calculate the product of numbers from 1 to 'n'.
def prod(n):
    # Initialize a variable 'result' to store the product, starting with 1.
    result = 1
    
    # Iterate through numbers from 1 to 'n' (inclusive).
    for i in range(1, n + 1):
        # Multiply 'result' by the current number 'i'.
        result *= i
    
    # Return the calculated product.
    return result

# Check if this script is the main module being run.
if __name__ == "__main__":
    # Read an integer 'n' from the user.
    n = int(input("Enter the value of 'n': "))
    
    # Call the 'prod' function to calculate the product.
    product = prod(n)
    
    # Print the calculated product with a formatted message.
    print(f"The product of numbers from 1 to {n} is: {product}")

Enter the value of 'n':  5


The product of numbers from 1 to 5 is: 120


## 3. Functions
a. Read ‘gpa.py’ and then ‘gpa2.py’. What does ‘gpa2.py’ do?

In [7]:
#gpa.py

gpa = int(input("What is your GPA? "))
if gpa > 2.0:
    print ("Your application was accepted.")
else:
    print ("You loser, get out of here.")

What is your GPA?  4


Your application was accepted.


In [46]:
def gpaResults(grade):
    if grade > 2.0:
        passDegree = True
    else:
        passDegree = False
    return passDegree

gpa = float(input("What is your GPA? "))
i = 10

if gpaResults(gpa):
    print("Your application was accepted.")
else:
    print("You loser, get out of here.")

print(i)

What is your GPA?  4


Your application was accepted.
10


The code prompts the user for their GPA, converts it to a floating-point number, and then determines whether their application was accepted based on whether the GPA is greater than 2.0, finally printing a corresponding message. 

## b. Write a function inside ‘gpa2.py’ that prints the letter grade given a number grade, thus a grade >=70 corresponds to ‘A’, a grade between 65 and 69 corresponds to ‘A-‘, 60-64 corresponds to B+, a grade 50-59 corresponds to B, a grade 40-49 is a C, and anything below 40 is an F.

In [54]:
# Function to determine if a student passes the degree based on GPA.
def gpaResults(grade):
    if grade >= 2.0:
        passDegree = True
    else:
        passDegree = False
    return passDegree

# Function to convert the input grade to a 100-point scale.
def convertTo100Scale(grade):
    # Assuming the input grade is in the range 1-4, you can convert it to a 100-point scale by multiplying by 25.
    converted_grade = grade * 25
    return converted_grade

# Function to print the letter grade based on the converted grade.
def printLetterGrade(grade):
    if grade >= 70:
        letter_grade = 'A'
    elif grade >= 65:
        letter_grade = 'A-'
    elif grade >= 60:
        letter_grade = 'B+'
    elif grade >= 50:
        letter_grade = 'B'
    elif grade >= 40:
        letter_grade = 'C'
    else:
        letter_grade = 'F'
    print(f"Your letter grade is: {letter_grade}")

# Get the input grade from the user (assumed to be in the range 1-4).
grade = float(input("Enter your grade (1-4): "))

# Convert the input grade to a 100-point scale.
converted_grade = convertTo100Scale(grade)

# Print the original grade and the converted grade.
print(f"Original grade: {grade}")
print(f"Grade in 100-point scale: {converted_grade}")

# Check if the student passes the degree based on the converted grade and print the letter grade.
if gpaResults(converted_grade):
    print("Your application was accepted.")
    printLetterGrade(converted_grade)
else:
    print("You loser, get out of here.")
    printLetterGrade(converted_grade)

Enter your grade (1-4):  2.3


Original grade: 2.3
Grade in 100-point scale: 57.49999999999999
Your application was accepted.
Your letter grade is: B


## c. Write a function to reverse a list. The function will be called ‘reverseList’. It will take as input a list, and returns the reversed list. Hint: use ‘len(list)’, and a while loop.

In [49]:
def reverseList(input_list):
    # Create an empty list to store the reversed elements
    reversed_list = []

    # Get the index of the last element in the input list
    index = len(input_list) - 1

    # Iterate through the input list in reverse order using a while loop
    while index >= 0:
        # Append the current element to the reversed_list
        reversed_list.append(input_list[index])
        
        # Move to the previous element (decrement the index)
        index -= 1

    # Return the reversed list
    return reversed_list

# Example usage:
original_list = [1, 2, 3, 4, 5]
reversed_result = reverseList(original_list)
print(reversed_result)

[5, 4, 3, 2, 1]


## d. Write a function to discover whether a string is a palindrome. It will return True or False. The function will be called ‘palindromeString (list)’. For example, ‘Anna’ is a palindrome.

In [50]:
def palindromeString(input_string):
    # Remove spaces and convert the string to lowercase for a case-insensitive check
    cleaned_string = input_string.replace(" ", "").lower()

    # Check if the cleaned string is equal to its reverse
    return cleaned_string == cleaned_string[::-1]

# Example usage:
input_str = "Anna"
result = palindromeString(input_str)
print(result)  # This will print True for "Anna"

True


## 4. Dictionaries

a. Can you use a dictionary to convert triplets of amino-acids to proteins? (see previous exercise), and see also: https://www.britannica.com/science/genetic-code that explains the conversion). The key should be a triplet and the value the amino-acid. e.g. the dictionary could be named: ‘dna2aa’. And dna2aa[‘aaa’] should be phenylamine

In [43]:
# Create a dictionary 'dna2aa' that maps DNA triplets to amino acids.
dna2aa = {
    'TTT': 'Phenylalanine',
    'TTC': 'Phenylalanine',
    'TTA': 'Leucine',
    'TTG': 'Leucine',
    'CTT': 'Leucine',
    'CTC': 'Leucine',
    'CTA': 'Leucine',
    'CTG': 'Leucine',
    'ATT': 'Isoleucine',
    'ATC': 'Isoleucine',
    'ATA': 'Isoleucine',
    'ATG': 'Methionine (Start)'
}

# Accessing a value in the dictionary using a DNA triplet as a key.
print(dna2aa['TTT'])  # This will print 'Phenylalanine'

Phenylalanine


b. Convert the two lists into a dictionary:
e.g. given: keys = ['Ten', 'Twenty', 'Thirty'], values = [10, 20, 30]

Expected output:
     {'Ten': 10, 'Twenty': 20, 'Thirty': 30}

In [42]:
# Define two lists, 'keys' and 'values'.
keys = ['Ten', 'Twenty', 'Thirty']
values = [10, 20, 30]

# Use the zip function to combine 'keys' and 'values' into a dictionary.
result_dict = dict(zip(keys, values))

# Print the resulting dictionary.
print(result_dict)

{'Ten': 10, 'Twenty': 20, 'Thirty': 30}


## 5. Numbers
a. Print the first 100 integers, starting from 0

In [41]:
# Loop through numbers from 0 to 99 (100 numbers in total).
for i in range(100):
    # Print the current value of 'i'.
    print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99


## b. Print the following numbers

1 

1 2

1 2 3

1 2 3 4

1 2 3 4 5

In [40]:
# Outer loop for rows, ranging from 1 to 5.
for i in range(1, 6):
    # Inner loop for numbers in each row, ranging from 1 to the current row number.
    for j in range(1, i + 1):
        # Print the value of 'j' followed by a space, without moving to a new line.
        print(j, end=' ')
    
    # Print a new line to move to the next row.
    print()

1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


## c. Write a function addSub() such that it can accept two numbers and calculate their sum and difference. It must also return them, e.g.
def addSub(a, b):

    #Your Code
res = addSub(5, 2)

print(res)


In [39]:
# Define a function named addSub that calculates the sum and difference of two numbers.
def addSub(a, b):
    # Calculate the sum of 'a' and 'b'.
    sum_result = a + b
    # Calculate the difference of 'a' and 'b'.
    diff_result = a - b
    # Return both the sum and difference as a tuple.
    return sum_result, diff_result

# Call the addSub function with arguments 5 and 2.
res = addSub(5, 2)
# The function returns a tuple with two values, so 'res' will contain (7, 3).
print(res)

(7, 3)


## 6. Lists

a. Given a list of numbers, print only those numbers which are divisible of 5 (write a function name ‘divByNum’. e.g., divByNum([10, 20, 33, 46, 55], 5), you should produce as output:

10

20

55

In [38]:
# Define a function named divByNum that prints numbers in a list that are divisible by a given divisor.
def divByNum(numbers, divisor):
    # Iterate through the numbers in the input list.
    for num in numbers:
        # Check if the current number is divisible by the specified divisor.
        if num % divisor == 0:
            # If it is divisible, print the number.
            print(num)

# Example usage:
numbers = [10, 20, 33, 46, 55]
divisor = 5
# Call the divByNum function with the list of numbers and the divisor as arguments.
divByNum(numbers, divisor)
# This will print the numbers in the list [10, 20, 55] since they are divisible by 5.

10
20
55


## 7. Strings
a. Given a string, display only those characters which are present at an even index number.
For example given str = "Alice" a function named ‘evenStr’ should output
           ‘A’, ‘i’, ‘e’.

In [37]:
# Define a function named evenStr that extracts characters at even indices from a given input string.
def evenStr(input_str):
    # Initialize an empty string to store the even-indexed characters.
    even_chars = ""
    
    # Iterate through the characters in the input string along with their indices.
    for index, char in enumerate(input_str):
        # Check if the index is even (0, 2, 4, etc.).
        if index % 2 == 0:
            # If the index is even, add the character to the even_chars string.
            even_chars += char
    
    # Return the string containing even-indexed characters.
    return even_chars

# Example usage:
input_str = "Alice"
# Call the evenStr function with the input string "Alice."
result = evenStr(input_str)
# Print the result, which will be "Aie" in this case.
print(result)

Aie


## b. Given an input string, write a function (freqCounter) to count occurrences of all characters within a string, e.g. str1 = "Apple" Expected Outcome:

{'A': 1, 'p': 2, 'l': 1, 'e': 1}

In [29]:
def freqCounter(input_str):
    # Initialize an empty dictionary to store character frequencies.
    char_count = {}
    
    # Iterate through the characters in the input string.
    for char in input_str:
        # Check if the character is already in the dictionary.
        if char in char_count:
            # If it is, increment the count for that character.
            char_count[char] += 1
        else:
            # If it's not, add it to the dictionary with a count of 1.
            char_count[char] = 1
    
    return char_count

# Example usage:
input_str = "Apple"
result = freqCounter(input_str)
print(result)

{'A': 1, 'p': 2, 'l': 1, 'e': 1}


## c. Reverse a string with a function that you should write a function named: revStr

e.g.: str1 = "Aloha" → aholA

In [32]:
def revStr(input_str):
    # Use string slicing to reverse the input string.
    reversed_str = input_str[::-1]
    return reversed_str

input_str = "Aloha"
result = revStr(input_str)
print(result)

aholA


## d. In a given string replace each punctuation symbol with #

e.g. given

str1 = '/*Jon is @developer & musician!!'

produce as output

##Jon is #developer # musician##

In [34]:
import string

def replacePunctuation(input_str, replacement_char='#'):
    # Create a translation table to map punctuation symbols to the replacement character.
    translation_table = str.maketrans(string.punctuation, replacement_char * len(string.punctuation))
    
    # Use the translate method to replace punctuation symbols in the input string.
    modified_str = input_str.translate(translation_table)
    
    return modified_str

# Example usage:
input_str = '/*Jon is @developer & musician!!'
result = replacePunctuation(input_str)
print(result)

##Jon is #developer # musician##
